### PR TITLE
CLOUD-526 Document the Operator API

### PR DESCRIPTION
**Pull request by fiowro**
[CLOUD-526](https://jira.percona.com/browse/CLOUD-526)

### CODE REVIEW

**Comment by fiowro** [comment]
screenshot

File: source/assets/code/api-create-cluster-request-json.txt [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+		    }
+		 },
+		 "podDisruptionBudget":{
+		    "maxUnavailable":1
+		 },
+		 "gracePeriod":30
+	      },
+	      "pmm":{
+		 "enabled":false,
+		 "image":"percona/percona-xtradb-cluster-operator:1.4.0-pmm",
```
Comment (hors) [inline source/assets/code/api-create-cluster-request-json.txt:67]: I think we need to use the latest images.

File: source/api.rst [hunk 2]
```diff
@@ -0,0 +1,613 @@
...
+Prerequisites
+-------------
+
+1. prepare
+
+   .. code-block:: yaml
+
+      # set correct API adress
+      KUBE_CLUSTER=$(kubectl config view --minify -o jsonpath='{.clusters[0].name}')
+      API_SERVER=$(kubectl config view -o jsonpath="{.clusters[?(@.name==\"$KUBE_CLUSTER\")].cluster.server}")
```
Comment (hors) [inline source/api.rst:45]: this var contains `https://IP` so we need to remove `https://` e.g. using the following `API_SERVER=$(kubectl config view -o jsonpath="{.clusters[?(@.name==\"$KUBE_CLUSTER\")].cluster.server}" | tr -d 'https://')` and fix line https://github.com/percona/percona-xtradb-cluster-operator/pull/518/files#diff-cc08f3ae0f8900da083eec2e9c304d71R56 or we need to fix all document by removing `https:// `  from e.g. https://github.com/percona/percona-xtradb-cluster-operator/pull/518/files#diff-cc08f3ae0f8900da083eec2e9c304d71R77

File: source/api.rst [hunk 3]
```diff
@@ -0,0 +1,613 @@
...
+      # create service account and get token
+      kubectl apply -f deploy/crd.yaml -f deploy/rbac.yaml
+      KUBE_TOKEN=$(kubectl get secret $(kubectl get serviceaccount percona-xtradb-cluster-operator -o jsonpath='{.secrets[0].name}') -o jsonpath='{.data.token}' | base64 --decode )
+
+2. get a list of PXC clusters
+
+   .. code-block:: bash
+
+      # run "kubectl --v=9 get pxc" command find curl command in the output
+      curl -k -XGET  -H "Authorization: Bearer $KUBE_TOKEN" "$APISERVER/apis/pxc.percona.com/v1/namespaces/default/perconaxtradbclusters?limit=500" | python -mjson.tool
```
Comment (hors) [inline source/api.rst:56]:
```suggestion
      curl -k -XGET  -H "Authorization: Bearer $KUBE_TOKEN" "$API_SERVER/apis/pxc.percona.com/v1/namespaces/default/perconaxtradbclusters?limit=500" | python -mjson.tool
```

File: source/api.rst [hunk 4]
```diff
@@ -0,0 +1,613 @@
...
+-------------
+
+1. prepare
+
+   .. code-block:: yaml
+
+      # set correct API adress
+      KUBE_CLUSTER=$(kubectl config view --minify -o jsonpath='{.clusters[0].name}')
+      API_SERVER=$(kubectl config view -o jsonpath="{.clusters[?(@.name==\"$KUBE_CLUSTER\")].cluster.server}")
+
```
Comment (hors) [inline source/api.rst:46]: As I can see for now we work with default namespace but we need to show how to work with different namespaces . So I suggest adding the following:
```
NAMESPACE='pxc'
kubectl create namespace $NAMESPACE
```

File: source/api.rst [hunk 5]
```diff
@@ -0,0 +1,613 @@
...
+      # create service account and get token
+      kubectl apply -f deploy/crd.yaml -f deploy/rbac.yaml
```
Comment (hors) [inline source/api.rst:48]:
```suggestion
      kubectl apply -f deploy/crd.yaml -f deploy/rbac.yaml -n $NAMESPACE
```

File: source/api.rst [hunk 6]
```diff
@@ -0,0 +1,613 @@
...
+      KUBE_TOKEN=$(kubectl get secret $(kubectl get serviceaccount percona-xtradb-cluster-operator -o jsonpath='{.secrets[0].name}') -o jsonpath='{.data.token}' | base64 --decode )
```
Comment (hors) [inline source/api.rst:49]:
```suggestion
      KUBE_TOKEN=$(kubectl get secret $(kubectl get serviceaccount percona-xtradb-cluster-operator -o jsonpath='{.secrets[0].name}' -n $NAMESPACE) -o jsonpath='{.data.token}' -n $NAMESPACE | base64 --decode )
```

File: source/api.rst [hunk 7]
```diff
@@ -0,0 +1,612 @@
...
+
+      # create service account and get token
+      kubectl apply -f deploy/crd.yaml -f deploy/rbac.yaml
+      KUBE_TOKEN=$(kubectl get secret $(kubectl get serviceaccount percona-xtradb-cluster-operator -o jsonpath='{.secrets[0].name}' -n $NAMESPACE) -o jsonpath='{.data.token}' -n $NAMESPACE | base64 --decode )
+
+2. get a list of PXC clusters
+
+   .. code-block:: bash
+
+      # run "kubectl --v=9 get pxc" command find curl command in the output
```
Comment (hors) [inline source/api.rst:55]: Kubectl Command:

```suggestion
      kubectl --v=9 get pxc
```

I think we can remove `--v=9` from the commands. It was just an example for developers how to get 'curl' ... @delgod  what do you think?

File: source/api.rst (hunk 2 above)
Comment (hors) [reply source/api.rst:58]: As I can see It was fixed for psmdb but not for pxc  doc :)

File: source/api.rst [hunk 8]
```diff
@@ -0,0 +1,626 @@
...
+.. code-block:: bash
+
+   Change the image of pxc containers inside the cluster
+
+**Kubectl Command:**
+
+.. code-block:: bash
+
+   kubectl --v=9 patch pxc cluster1 --type=merge --patch '{  
+   "spec": {"pxc":{ "image": "percona/percona-xtradb-cluster-operator:1.5.0-pxc5.7" }
```
Comment (hors) [inline source/api.rst:356]: as you know we do not have  such image starting from 1.5.0.  Please use `percona/percona-xtradb-cluster:5.7.30-31.43`

File: source/api.rst [hunk 9]
```diff
@@ -0,0 +1,626 @@
...
+**cURL Request:**
+
+.. code-block:: bash
+
+   curl -k -v -XPATCH 'https://$API_SERVER/apis/pxc.percona.com/v1/namespaces/default/perconaxtradbclusters/cluster1' \
+               -H "Authorization: Bearer $KUBE_TOKEN" \
+               -H "Accept: application/json" \
+               -H "Content-Type: application/merge-patch+json" 
+               -d '{  
+                 "spec": {"pxc":{ "image": "percona/percona-xtradb-cluster-operator:1.5.0-pxc5.7" }
```
Comment (hors) [inline source/api.rst:381]:
```suggestion
                 "spec": {"pxc":{ "image": "percona/percona-xtradb-cluster:5.7.30-31.43" }
```

File: source/assets/code/api-create-cluster-request-json.txt [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+	   },
+	   "spec":{
+	      "secretsName":"my-cluster-secrets",
+	      "vaultSecretName":"keyring-secret-vault",
+	      "sslSecretName":"my-cluster-ssl",
+	      "sslInternalSecretName":"my-cluster-ssl-internal",
+	      "allowUnsafeConfigurations":true,
+	      "pxc":{
+		 "size":3,
+		 "image":"percona/percona-xtradb-cluster-operator:1.5.0-pxc8.0",
```
Comment (hors) [inline source/assets/code/api-create-cluster-request-json.txt:20]:
```suggestion
		 "image":"percona/percona-xtradb-cluster:8.0.19-10.1",
```

File: source/assets/code/api-create-cluster-response-json.txt [hunk 11]
```diff
@@ -0,0 +1,375 @@
...
+		       }
+		    }
+		 }
+	      },
+	      "pxc":{
+		 "affinity":{
+		    "antiAffinityTopologyKey":"none"
+		 },
+		 "gracePeriod":600,
+		 "image":"percona/percona-xtradb-cluster-operator:1.5.0-pxc8.0",
```
Comment (hors) [inline source/assets/code/api-create-cluster-response-json.txt:351]:
```suggestion
		 "image":"percona/percona-xtradb-cluster:8.0.19-10.1",
```

File: source/assets/code/api-update-cluster-image-response-json.txt [hunk 12]
```diff
@@ -0,0 +1,595 @@
...
+		       }
+		    }
+		 }
+	      },
+	      "pxc":{
+		 "affinity":{
+		    "antiAffinityTopologyKey":"none"
+		 },
+		 "gracePeriod":600,
+		 "image":"percona/percona-xtradb-cluster-operator:1.5.0-pxc5.7",
```
Comment (hors) [inline source/assets/code/api-update-cluster-image-response-json.txt:528]:
```suggestion
		 "image":"percona/percona-xtradb-cluster:5.7.30-31.43",
```

File: source/assets/code/api-scale-cluster-response-json.txt [hunk 13]
```diff
@@ -0,0 +1,586 @@
...
...ket\":\"S3-BACKUP-BUCKET-NAME-HERE\",\"credentialsSecret\":\"my-cluster-name-backup-s3\",\"region\":\"us-west-2\"},\"type\":\"s3\"}}},\"pmm\":{\"enabled\":false,\"image\":\"percona/percona-xtradb-cluster-operator:1.5.0-pmm\",\"serverHost\":\"monitoring-service\",\"serverUser\":\"pmm\"},\"proxysql\":{\"affinity\":{\"antiAffinityTopologyKey\":\"none\"},\"enabled\":true,\"gracePeriod\":30,\"image\":\"percona/percona-xtradb-cluster-operator:1.5.0-proxysql\",\"podDisruptionBudget\":{\"maxUnavailable\":1},\"resources\":{\"requests\":null},\"size\":3,\"volumeSpec\":{\"persistentVolumeClaim\":{\"resources\":{\"requests\":{\"storage\":\"2Gi\"}}}}},\"pxc\":{\"affinity\":{\"antiAffinityTopologyKey\":\"none\"},\"gracePeriod\":600,\"image\":\"percona/percona-xtradb-cluster-operator:1.5.0-pxc8.0\",\"podDisruptionBudget\":{\"maxUnavailable\":1},\"resources\":{\"requests\":null},\"size\":3,\"volumeSpec\":{\"persistentVolumeClaim\":{\"resources\":{\"requests\":{\"storage\":\"6Gi\"}}}}},\"secretsName\":\"my-cluster-secrets\",\"sslInternalSecretName\":\"my-cluster-ssl-internal\",\"sslSecretName\":\"my-cluster-ssl\",\"updateStrategy\":\"RollingUpdate\",\"vaultSecretName\":\"keyring-secret-vault\"}}\n"
```
Comment (hors) [inline source/assets/code/api-scale-cluster-response-json.txt:8]:
```suggestion
		 "kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"pxc.percona.com/v1-5-0\",\"kind\":\"PerconaXtraDBCluster\",\"metadata\":{\"annotations\":{},\"finalizers\":[\"delete-pxc-pods-in-order\"],\"name\":\"cluster1\",\"namespace\":\"default\"},\"spec\":{\"allowUnsafeConfigurations\":true,\"backup\":{\"image\":\"percona/percona-xtradb-cluster-operator:1.5.0-pxc8.0-backup\",\"schedule\":[{\"keep\":3,\"name\":\"sat-night-backup\",\"schedule\":\"0 0 * * 6\",\"storageName\":\"s3-us-west\"},{\"keep\":5,\"name\":\"daily-backup\",\"schedule\":\"0 0 * * *\",\"storageName\":\"fs-pvc\"}],\"serviceAccountName\":\"percona-xtradb-cluster-operator\",\"storages\":{\"fs-pvc\":{\"type\":\"filesystem\",\"volume\":{\"persistentVolumeClaim\":{\"accessModes\":[\"ReadWriteOnce\"],\"resources\":{\"requests\":{\"storage\":\"6Gi\"}}}}},\"s3-us-west\":{\"s3\":{\"bucket\":\"S3-BACKUP-BUCKET-NAME-HERE\",\"credentialsSecret\":\"my-cluster-name-backup-s3\",\"region\":\"us-west-2\"},\"type\":\"s3\"}}},\"pmm\":{\"enabled\":false,\"image\":\"percona/percona-xtradb-cluster-operator:1.5.0-pmm\",\"serverHost\":\"monitoring-service\",\"serverUser\":\"pmm\"},\"proxysql\":{\"affinity\":{\"antiAffinityTopologyKey\":\"none\"},\"enabled\":true,\"gracePeriod\":30,\"image\":\"percona/percona-xtradb-cluster-operator:1.5.0-proxysql\",\"podDisruptionBudget\":{\"maxUnavailable\":1},\"resources\":{\"requests\":null},\"size\":3,\"volumeSpec\":{\"persistentVolumeClaim\":{\"resources\":{\"requests\":{\"storage\":\"2Gi\"}}}}},\"pxc\":{\"affinity\":{\"antiAffinityTopologyKey\":\"none\"},\"gracePeriod\":600,\"image\":\"percona/percona-xtradb-cluster:8.0.19-10.1\",\"podDisruptionBudget\":{\"maxUnavailable\":1},\"resources\":{\"requests\":null},\"size\":3,\"volumeSpec\":{\"persistentVolumeClaim\":{\"resources\":{\"requests\":{\"storage\":\"6Gi\"}}}}},\"secretsName\":\"my-cluster-secrets\",\"sslInternalSecretName\":\"my-cluster-ssl-internal\",\"sslSecretName\":\"my-cluster-ssl\",\"updateStrategy\":\"RollingUpdate\",\"vaultSecretName\":\"keyring-secret-vault\"}}\n"
```

File: source/assets/code/api-update-cluster-image-response-json.txt [hunk 14]
```diff
@@ -0,0 +1,595 @@
...
...ket\":\"S3-BACKUP-BUCKET-NAME-HERE\",\"credentialsSecret\":\"my-cluster-name-backup-s3\",\"region\":\"us-west-2\"},\"type\":\"s3\"}}},\"pmm\":{\"enabled\":false,\"image\":\"percona/percona-xtradb-cluster-operator:1.5.0-pmm\",\"serverHost\":\"monitoring-service\",\"serverUser\":\"pmm\"},\"proxysql\":{\"affinity\":{\"antiAffinityTopologyKey\":\"none\"},\"enabled\":true,\"gracePeriod\":30,\"image\":\"percona/percona-xtradb-cluster-operator:1.5.0-proxysql\",\"podDisruptionBudget\":{\"maxUnavailable\":1},\"resources\":{\"requests\":null},\"size\":3,\"volumeSpec\":{\"persistentVolumeClaim\":{\"resources\":{\"requests\":{\"storage\":\"2Gi\"}}}}},\"pxc\":{\"affinity\":{\"antiAffinityTopologyKey\":\"none\"},\"gracePeriod\":600,\"image\":\"percona/percona-xtradb-cluster-operator:1.5.0-pxc8.0\",\"podDisruptionBudget\":{\"maxUnavailable\":1},\"resources\":{\"requests\":null},\"size\":3,\"volumeSpec\":{\"persistentVolumeClaim\":{\"resources\":{\"requests\":{\"storage\":\"6Gi\"}}}}},\"secretsName\":\"my-cluster-secrets\",\"sslInternalSecretName\":\"my-cluster-ssl-internal\",\"sslSecretName\":\"my-cluster-ssl\",\"updateStrategy\":\"RollingUpdate\",\"vaultSecretName\":\"keyring-secret-vault\"}}\n"
```
Comment (hors) [inline source/assets/code/api-update-cluster-image-response-json.txt:8]:
```suggestion
		 "kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"pxc.percona.com/v1-5-0\",\"kind\":\"PerconaXtraDBCluster\",\"metadata\":{\"annotations\":{},\"finalizers\":[\"delete-pxc-pods-in-order\"],\"name\":\"cluster1\",\"namespace\":\"default\"},\"spec\":{\"allowUnsafeConfigurations\":true,\"backup\":{\"image\":\"percona/percona-xtradb-cluster-operator:1.5.0-pxc8.0-backup\",\"schedule\":[{\"keep\":3,\"name\":\"sat-night-backup\",\"schedule\":\"0 0 * * 6\",\"storageName\":\"s3-us-west\"},{\"keep\":5,\"name\":\"daily-backup\",\"schedule\":\"0 0 * * *\",\"storageName\":\"fs-pvc\"}],\"serviceAccountName\":\"percona-xtradb-cluster-operator\",\"storages\":{\"fs-pvc\":{\"type\":\"filesystem\",\"volume\":{\"persistentVolumeClaim\":{\"accessModes\":[\"ReadWriteOnce\"],\"resources\":{\"requests\":{\"storage\":\"6Gi\"}}}}},\"s3-us-west\":{\"s3\":{\"bucket\":\"S3-BACKUP-BUCKET-NAME-HERE\",\"credentialsSecret\":\"my-cluster-name-backup-s3\",\"region\":\"us-west-2\"},\"type\":\"s3\"}}},\"pmm\":{\"enabled\":false,\"image\":\"percona/percona-xtradb-cluster-operator:1.5.0-pmm\",\"serverHost\":\"monitoring-service\",\"serverUser\":\"pmm\"},\"proxysql\":{\"affinity\":{\"antiAffinityTopologyKey\":\"none\"},\"enabled\":true,\"gracePeriod\":30,\"image\":\"percona/percona-xtradb-cluster-operator:1.5.0-proxysql\",\"podDisruptionBudget\":{\"maxUnavailable\":1},\"resources\":{\"requests\":null},\"size\":3,\"volumeSpec\":{\"persistentVolumeClaim\":{\"resources\":{\"requests\":{\"storage\":\"2Gi\"}}}}},\"pxc\":{\"affinity\":{\"antiAffinityTopologyKey\":\"none\"},\"gracePeriod\":600,\"image\":\"percona/percona-xtradb-cluster:8.0.19-10.1\",\"podDisruptionBudget\":{\"maxUnavailable\":1},\"resources\":{\"requests\":null},\"size\":3,\"volumeSpec\":{\"persistentVolumeClaim\":{\"resources\":{\"requests\":{\"storage\":\"6Gi\"}}}}},\"secretsName\":\"my-cluster-secrets\",\"sslInternalSecretName\":\"my-cluster-ssl-internal\",\"sslSecretName\":\"my-cluster-ssl\",\"updateStrategy\":\"RollingUpdate\",\"vaultSecretName\":\"keyring-secret-vault\"}}\n"
```

File: source/assets/code/api-scale-cluster-response-json.txt [hunk 15]
```diff
@@ -0,0 +1,586 @@
...
+		       }
+		    }
+		 }
+	      },
+	      "pxc":{
+		 "affinity":{
+		    "antiAffinityTopologyKey":"none"
+		 },
+		 "gracePeriod":600,
+		 "image":"percona/percona-xtradb-cluster-operator:1.5.0-pxc5.7",
```
Comment (hors) [inline source/assets/code/api-scale-cluster-response-json.txt:522]:
```suggestion
		 "image":"percona/percona-xtradb-cluster:5.7.30-31.43",
```

File: source/assets/code/api-update-cluster-image-request-json.txt [hunk 16]
```diff
@@ -0,0 +1,5 @@
+.. code-block:: json
+
+	{
+	"spec": {"pxc":{ "image": "percona/percona-xtradb-cluster-operator:1.5.0-pxc5.7" }
```
Comment (hors) [inline source/assets/code/api-update-cluster-image-request-json.txt:4]:
```suggestion
	"spec": {"pxc":{ "image": "percona/percona-xtradb-cluster:5.7.30-31.43" }
```

File: source/api.rst [hunk 17]
```diff
@@ -0,0 +1,626 @@
...
+      kubectl apply -f deploy/crd.yaml -f deploy/rbac.yaml
+      KUBE_TOKEN=$(kubectl get secret $(kubectl get serviceaccount percona-xtradb-cluster-operator -o jsonpath='{.secrets[0].name}' -n default) -o jsonpath='{.data.token}' -n default | base64 --decode )
+      
+
+3. Get a list of PXC clusters
+
+   .. code-block:: bash
+
+      kubectl --v=9 get pxc
+      curl -k -XGET  -H "Authorization: Bearer $KUBE_TOKEN" "$API_SERVER/apis/pxc.percona.com/v1/namespaces/default/perconaxtradbclusters?limit=500" | python -mjson.tool
```
Comment (hors) [inline source/api.rst:70]:
```suggestion
      curl -k -XGET  -H "Authorization: Bearer $KUBE_TOKEN" "https://$API_SERVER/apis/pxc.percona.com/v1/namespaces/default/perconaxtradbclusters?limit=500" | python -mjson.tool
```

File: source/api.rst [hunk 18]
```diff
@@ -0,0 +1,626 @@
...
+      Substitute ``default`` with your namespace name if you use a different
+      one.
+
+2. Prepare
+
+   .. code-block:: yaml
+
+      # set correct API address
+      KUBE_CLUSTER=$(kubectl config view --minify -o jsonpath='{.clusters[0].name}')
+      API_SERVER=$(kubectl config view -o jsonpath="{.clusters[?(@.name==\"$KUBE_CLUSTER\")].cluster.server}")
```
Comment (hors) [inline source/api.rst:58]:
```suggestion
      API_SERVER=$(kubectl config view -o jsonpath="{.clusters[?(@.name==\"$KUBE_CLUSTER\")].cluster.server}" | tr -d 'https://')
```

File: source/api.rst [hunk 19]
```diff
@@ -0,0 +1,626 @@
...
+      kubectl apply -f deploy/crd.yaml -f deploy/rbac.yaml -n default
+      KUBE_TOKEN=$(kubectl get secret $(kubectl get serviceaccount percona-xtradb-cluster-operator -o jsonpath='{.secrets[0].name}' -n default) -o jsonpath='{.data.token}' -n default | base64 --decode )
+      
+
+3. Get a list of PXC clusters
+
+   .. code-block:: bash
+
+      kubectl --v=9 get pxc
+      curl -k -XGET  -H "Authorization: Bearer $KUBE_TOKEN" "https://$API_SERVER/apis/pxc.percona.com/v1/namespaces/default/perconaxtradbclusters?limit=500" | python -mjson.tool
```
Comment (delgod) [inline source/api.rst:70]:
```suggestion
3. Check if we have correct values in API_SERVER and KUBE_TOKEN variabled

   .. code-block:: bash

      curl -k -XGET  -H "Authorization: Bearer $KUBE_TOKEN" "https://$API_SERVER/apis/pxc.percona.com/v1/namespaces/default/perconaxtradbclusters?limit=500" | python -mjson.tool
```

**Review by delgod — CHANGES_REQUESTED** [review]

File: source/api.rst (hunk 7 above)
Comment (delgod) [reply source/api.rst]: have sense

File: source/api.rst [hunk 20]
```diff
@@ -0,0 +1,626 @@
...
+      kubectl apply -f deploy/crd.yaml -f deploy/rbac.yaml -n default
+      KUBE_TOKEN=$(kubectl get secret $(kubectl get serviceaccount percona-xtradb-cluster-operator -o jsonpath='{.secrets[0].name}' -n default) -o jsonpath='{.data.token}' -n default | base64 --decode )
+      
+
+3. Get a list of PXC clusters
+
+   .. code-block:: bash
+
+      kubectl get pxc
+      curl -k -XGET  -H "Authorization: Bearer $KUBE_TOKEN" "https://$API_SERVER/apis/pxc.percona.com/v1/namespaces/default/perconaxtradbclusters?limit=500" | python -mjson.tool
```
Comment (delgod) [inline source/api.rst:70]: we should remove these lines or rework in the following way
```suggestion
3. Check if we have correct values in API_SERVER and KUBE_TOKEN variables

   .. code-block:: bash

      curl -k -XGET  -H "Authorization: Bearer $KUBE_TOKEN" "https://$API_SERVER/apis/pxc.percona.com/v1/namespaces/default/perconaxtradbclusters?limit=500" | python -mjson.tool
      
      expected output is
    
     .. code-block:: bash

    {{...@fiowro, please PUT REAL OUTPUT HERE...}}
```

**Review by delgod — CHANGES_REQUESTED** [review]

**Review by delgod — APPROVED** [review]

**Pull request merged**